docs/0001-teleport-job-worker.md

```diff
@@ -0,0 +1,185 @@
+as show below:
+
+```
+  ┌───────────────────────────┐
+  │                           │
+ready      ┌────►stopped◄─────┤
+  │        │                  │
+  │        │                  │
+  │        ├────►failed◄──────┘
+  │        │                   
+  ▼        │                   
+running────┴────►succeeded     
+```
+
+Resource control will be implemented using cgroups. A parent directory `job-worker` will be created
+under `/sys/fs/cgroup` that will contain a directory for every job created. For example, the path for `job-a` would look
+like `/sys/fs/cgroup/job-worker/job-a`. Under this path, the following files would be updated to manage the resources
+for the job:
+
+* `cgroup.procs` contains the PID of the job, i.e. the forked process; cgroups will be applied to this PID in order to
+  control the resources of all child processes
+* `cpu.max` limits CPU usage
+* `memory.max` limits memory usage
+* `io.max` limits disk IO
+
+Note that for this implementation, job metadata will remain in memory until the application exits. A real implementation
+of this service would store job metadata in some kind of persistent storage like a database.
+
+### Proto Specification
+
+The service will implement the following protobufs:
+
+* [job](../api/proto/job/job.proto)
+
+### API
+
+The service's API will contain the resources listed below and be made available using gRPC.
+
+#### job
+
+A job is responsible for managing the lifecycle of the process given to it.
+
+* `Start` start a new job and begin execution of the specified command immediately
+* `Stop` stop execution of the specified job immediately
+* `Query` query details about specified job; this function can run on a job of any status
+* `Output` get the full output (stdout and stderr) of any existing job; the output from the job will be written to
+  memory as the process executes so that it can be retrieved by calls to this function; for this exercise, output will
+  remain available until the job worker has exited; for real world uses, the output would need to be removed or written
+  to persistent storage because of memory limitations
+
+Details of a job's functions are listed in [job.proto](../api/proto/job/job.proto).
+
+### Library Godoc
+
+```
+package library // import "github.com/kurczynski/teleport-job-worker/pkg/library"
+
+TYPES
+
+// Job Contains information to interact with jobs.
+type Job struct {
+	// Has unexported fields.
+}
+
+// GetJob Return the job that has the specified ID; return nil if no job exists.
+func GetJob(id string) *Job
+
+// NewJob Create a new job to run the specified command.
+func NewJob(command string, args ...string) *Job
+
+// ID Returns the ID of the job.
+func (j *Job) ID() string
+
+// Output Get the full output (stdout and stderr) from the job.
+func (j *Job) Output() *job.OutputRequest
+
+// Query Get information about the job.
+func (j *Job) Query() *job.Info
```

Comment:
Library should be separate from the API server / not coupled to gRPC.

(Assuming `job` is a generated protobuf package.)

Reply:
Good thinking, I'll update the library section.